ResearchKit/Common/ORKPasscodeStepViewController.m

```diff
@@ -0,0 +1,571 @@
+            break;
+            
+        case ORKPasscodeStateSaved:
+            _passcodeStepView.headerView.captionLabel.text = ORKLocalizedString(@"PASSCODE_SAVED_MESSAGE", nil);
+            _passcodeStepView.headerView.instructionLabel.text = @"";
+            _passcodeStepView.textField.hidden = YES;
+            [self makePasscodeViewResignFirstResponder];
+            break;
+            
+        case ORKPasscodeStateOldEntry:
+            _passcodeStepView.headerView.captionLabel.text = ORKLocalizedString(@"PASSCODE_OLD_ENTRY_MESSAGE", nil);
+            _position = 0;
+            _passcode = [NSMutableString new];
+            _confirmPasscode = [NSMutableString new];
+            break;
+            
+        case ORKPasscodeStateNewEntry:
+            _passcodeStepView.headerView.captionLabel.text = ORKLocalizedString(@"PASSCODE_NEW_ENTRY_MESSAGE", nil);
+            _position = 0;
+            _passcode = [NSMutableString new];
+            _confirmPasscode = [NSMutableString new];
+            break;
+            
+        case ORKPasscodeStateConfirmNewEntry:
+            _passcodeStepView.headerView.captionLabel.text = ORKLocalizedString(@"PASSCODE_CONFIRM_NEW_ENTRY_MESSAGE", nil);
+            _position = 0;
+            _confirmPasscode = [NSMutableString new];
+            break;
+    }
+    
+    // Update the textField's text.
+    NSString *text = (_passcodeStepView.passcodeType == ORKPasscodeType4Digit) ? k4DigitPin : k6DigitPin;
+    _passcodeStepView.textField.text = text;
+    
+    // Enable input.
+    _isChangingState = NO;
+}
+
+- (void)initializeInternalButtonItems {
+    [super initializeInternalButtonItems];
+    
+    self.internalContinueButtonItem = nil;
+    self.internalDoneButtonItem = nil;
+}
+                                                              
+- (void)showValidityAlertWithMessage:(NSString *)text {
+    
+    UIAlertController *alert = [UIAlertController alertControllerWithTitle:ORKLocalizedString(@"PASSCODE_INVALID_ALERT_TITLE", nil)
+                                                                   message:text
+                                                            preferredStyle:UIAlertControllerStyleAlert];
+    [alert addAction:[UIAlertAction actionWithTitle:ORKLocalizedString(@"BUTTON_OK", nil)
+                                              style:UIAlertActionStyleDefault
+                                            handler:nil]];
+    [self presentViewController:alert animated:YES completion:nil];
+}
+
+- (ORKStepResult *)result {
+    ORKStepResult *stepResult = [super result];
+    
+    ORKPasscodeResult *passcodeResult = [[ORKPasscodeResult alloc] initWithIdentifier:[self passcodeStep].identifier];
+    passcodeResult.passcodeSaved = _isPasscodeSaved;
+    
+    stepResult.results = @[passcodeResult];
+    return stepResult;
+}
+
+#pragma mark - Helpers
+
+- (void)cancelButtonAction {
+    [self.passcodeDelegate passcodeViewControllerDidCancel:self];
```

Comment:
Nice to check AGAIN whether `passcodeViewControllerDidCancel:` has been implemented. (For example, passcodeDelegate can be changed.)